### DISCUSSION TITLE
How to forward stream from wireguard vpn to rtsp-simple-server running in docker

### DISCUSSION THREAD

**Discussion by dbuscaino** [bluenviron/mediamtx#1495] [Questions] [answered]
Hi, I have a raspberry that streams the camera with ffmpeg to a remote server hosted in AWS in which rtsp-simple-server is running with docker.
I need docker to read the stream with webrtc when behind a NAT. Here my  docker command: 

`/usr/bin/docker run --restart=always -d -it -p 8189:8189/udp -p 8554:8554 -p 8889:8889 -v /home/ubuntu/rtsp-simple-server/rtsp-simple-server.yml:/rtsp-simple-server.yml -v /home/ubuntu/rtsp-simple-server/server.crt:/server.crt -v /home/ubuntu/rtsp-simple-server/server.key:/server.key aler9/rtsp-simple-server`

I want to send my stream with udp to reduce latency as much as possible because stable internet connection is not guaranteed. In a previous request you suggested that I have to create a VPN with wireguard. Unfortunately I'm not very experienced so I apologize for any inaccuracies that may be there.

Now I think I created the VPN correctly because I can ping and curl remote server via VPN tunnel but I can't figured out how to forward my stream to rtsp-simple-server. 

In particular I ask for suggestions on:
- How to modify my ffmpeg command to send the stream with udp (I'm not sure if using rtsp specifying from ffmpeg to use udp is enough)
- To which port of rtsp-simple-server I should send the stream 
- How set wireguard to forward stream to rtsp-simple-server when running with docker.

Here my setup:
- My current ffmpeg command to send rtsp stream with tcp is:
 `ffmpeg -input_format h264 -f video4linux2 -video_size 800x600 -framerate 30 -i /dev/video0 -c:v copy -an -preset ultrafast -tune zerolatency -rtsp_transport tcp -f rtsp://server_public_ip:8554/specific_path`

- My current wireguard configuration on the raspberry:  
```
[Interface] 
PrivateKey = peer_private_key
Address = 10.8.0.2/24

[Peer]
PublicKey = server_public_key
AllowedIPs = 10.8.0.0/24
Endpoint = server_public_ip:51820
```

- My current wireguard configuration on the public server:

```
[Interface]
Address = 10.8.0.1/24
SaveConfig = true
PreUp = sysctl -w net.ipv4.ip_forward=1
PostUp = ufw route allow in on wg0 out on ens3
PostUp = iptables -t nat -I POSTROUTING -o ens3 -j MASQUERADE
PreDown = ufw route delete allow in on wg0 out on ens3
PreDown = iptables -t nat -D POSTROUTING -o ens3 -j MASQUERADE
ListenPort = 51820
PrivateKey = server_private_key

[Peer]
PublicKey = peer_public_key
AllowedIPs = 10.8.0.2/32
Endpoint = 151.***.***.***:46972
```

I notice that if I use this ffmpeg command 
`ffmpeg -input_format h264 -f video4linux2 -video_size 800x600 -framerate 30 -i /dev/video0 -c:v copy -an -preset ultrafast -tune zerolatency -rtsp_transport udp -f rtsp://10.8.0.1`
to send the stream, from the server something is received:

```
$ sudo wg
interface: wg0
  public key: server_public_key
  private key: (hidden)
  listening port: 51820

peer: peer_public_key
  endpoint: 151.***.***.***:46972
  allowed ips: 10.8.0.2/32
  latest handshake: 19 seconds ago
  transfer: 628 B received, 124 B sent
```

I appreciate any kind of help! Thank you!

**Comment by aler9** [comment, marked as answer]
Hello,

> How to modify my ffmpeg command to send the stream with udp (I'm not sure if using rtsp specifying from ffmpeg to use udp is enough)

FFmpeg by default uses the UDP transport protocol of RTSP, so nothing should be done. Please also note that while Wireguard works with UDP on an outer level, it supports transfering any kind of layer 3 packets, like ICMP (ping), UDP, TCP and all that's below. They are all encrypted and wrapped into UDP packets in a transparent way. Therefore you can use all RTSP transport protocols (even multicast) without any limitation.

> To which port of rtsp-simple-server I should send the stream

The RTSP port is 8554 as usual:

```
ffmpeg -input_format h264 -f video4linux2 -video_size 800x600 -framerate 30 -i /dev/video0 -c:v copy -an -preset ultrafast -tune zerolatency -rtsp_transport udp -f rtsp://10.8.0.1:8554
```

> How set wireguard to forward stream to rtsp-simple-server when running with docker.

Use the `--network=host` flag. Docker and RTSP/UDP doesn't get along well without that flag:

```
/usr/bin/docker run --restart=always -d --network=host -v /home/ubuntu/rtsp-simple-server/rtsp-simple-server.yml:/rtsp-simple-server.yml -v /home/ubuntu/rtsp-simple-server/server.crt:/server.crt -v /home/ubuntu/rtsp-simple-server/server.key:/server.key aler9/rtsp-simple-server
```

Otherwise, you can disable the UDP transport protocol, by editing the server configuration:

```yml
protocols: [tcp]
```

[a comment by an automated account is not shown]
